contracts/oracle/FastPriceFeed.sol

```diff
@@ -236,80 +303,152 @@ contract FastPriceFeed is ISecondaryPriceFeed, IFastPriceFeed, Governable {
+        return fastPrice;
     }
 
-    function favorFastPrice() public view returns (bool) {
+    function favorFastPrice(address _token) public view returns (bool) {
         if (isSpreadEnabled) {
             return false;
         }
 
         if (disableFastPriceVoteCount >= minAuthorizations) {
+            // force a spread if watchers have flagged an issue with the fast price
+            return false;
+        }
+
+        (/* uint256 prevRefPrice */, /* uint256 refTime */, uint256 cumulativeRefDelta, uint256 cumulativeFastDelta) = getPriceData(_token);
+        if (cumulativeFastDelta > cumulativeRefDelta && cumulativeFastDelta.sub(cumulativeRefDelta) > maxCumulativeDeltaDiffs[_token]) {
+            // force a spread if the cumulative delta for the fast price feed exceeds the cumulative delta
+            // for the Chainlink price feed by the maxCumulativeDeltaDiff allowed
             return false;
         }
 
         return true;
     }
 
+    function getPriceData(address _token) public view returns (uint256, uint256, uint256, uint256) {
+        PriceDataItem memory data = priceData[_token];
+        return (uint256(data.refPrice), uint256(data.refTime), uint256(data.cumulativeRefDelta), uint256(data.cumulativeFastDelta));
+    }
+
     function _setPricesWithBits(uint256 _priceBits, uint256 _timestamp) private {
         bool shouldUpdate = _setLastUpdatedValues(_timestamp);
 
         if (shouldUpdate) {
             address _fastPriceEvents = fastPriceEvents;
+            address _vaultPriceFeed = vaultPriceFeed;
 
             for (uint256 j = 0; j < 8; j++) {
                 uint256 index = j;
                 if (index >= tokens.length) { return; }
 
                 uint256 startBit = 32 * j;
-                uint256 price = (_priceBits >> startBit) & PRICE_BITMASK;
+                uint256 price = (_priceBits >> startBit) & BITMASK_32;
 
                 address token = tokens[j];
                 uint256 tokenPrecision = tokenPrecisions[j];
                 uint256 adjustedPrice = price.mul(PRICE_PRECISION).div(tokenPrecision);
-                prices[token] = adjustedPrice;
 
-                _emitPriceEvent(_fastPriceEvents, token, adjustedPrice);
+                _setPrice(token, adjustedPrice, _vaultPriceFeed, _fastPriceEvents);
             }
         }
     }
 
+    function _setPrice(address _token, uint256 _price, address _vaultPriceFeed, address _fastPriceEvents) private {
+        if (_vaultPriceFeed != address(0)) {
+            uint256 refPrice = IVaultPriceFeed(_vaultPriceFeed).getLatestPrimaryPrice(_token);
+            uint256 fastPrice = prices[_token];
+
+            (uint256 prevRefPrice, uint256 refTime, uint256 cumulativeRefDelta, uint256 cumulativeFastDelta) = getPriceData(_token);
+
+            if (prevRefPrice > 0) {
+                uint256 refDeltaAmount = refPrice > prevRefPrice ? refPrice.sub(prevRefPrice) : prevRefPrice.sub(refPrice);
```

Comment:
btw why not store delta as a percent? otherwise we need to re-configure max diffs each time prices changes (eg ETH grew 1k -> 2k in last couple months)